acnportal/acnsim/network/charging_network.py

```diff
@@ -163,22 +228,88 @@ def update_pilots(self, pilots, i, period):
         Station IDs not registered in the network are silently ignored.
 
         Args:
-            pilots (Dict[str, List[number]]): Dictionary mapping station_ids to lists of pilot signals. Each index in
-                the array corresponds to an a period of the simulation. [A]
+            pilots (np.Array): numpy array with a row for each station_id
+                and a column for each time. Each entry in the Array corresponds to
+                a charging rate (in A) at the staion given by the row at a time given by the column.
             i (int): Current time index of the simulation.
             period (float): Length of the charging period. [minutes]
 
         Returns:
             None
         """
-        for station_id in self._EVSEs:
-            if station_id in pilots and i < len(pilots[station_id]):
-                new_rate = pilots[station_id][i]
-            else:
-                new_rate = 0
-            self._EVSEs[station_id].set_pilot(new_rate, self._voltages[station_id], period)
+        ids = self.station_ids
+        for station_number in range(len(ids)):
+            new_rate = pilots[station_number, i]
+            self._EVSEs[ids[station_number]].set_pilot(new_rate, self._voltages[station_number], period)
+
+    def constraint_current(self, schedule_matrix, constraints=None, time_indices=None, linear=False):
+        """ Return the aggregate currents subject to the given constraints. If constraints=None,
+        return all aggregate currents.
+
+        Args:
+            schedule_matrix (np.Array): 2-D matrix with each row corresponding to an EVSE and each
+                column corresponding to a time index in the schedule.
+            constraints (List[str]): List of constraint id's for which to calculate aggregate current. If
+                None, calculates aggregate currents for all constraints.
+            time_indices (List[int]): List of time indices for which to calculate aggregate current. If None, 
+                calculates aggregate currents for all timesteps.
+            linear (bool): If True, linearize all constraints to a more conservative but easier to compute constraint by
+                ignoring the phase angle and taking the absolute value of all load coefficients. Default False.
+
+        Returns:
+            List[complex]: Aggregate currents subject to the given constraints.
+        """
+        # Convert list of constraint id's to list of indices in constraint matrix
+        if constraints is not None:
+            constraint_indices = [i for i in range(len(self.constraint_index)) if self.constraint_index[i] in constraints]
+        else:
+            constraint_indices = list(range(len(self.constraint_index)))
+
+        # If we only want the constraint currents at specific time indices,
+        # index schedule_matrix columns using these indices
+        if time_indices is not None:
+            schedule_matrix = schedule_matrix[:, time_indices]
```

Comment:
This will (I think) overwrite the original schedule_matrix. Use a different variable name or make a copy.